### ISSUE TITLE
tls-alpn-01 challenge validation fails with Traefik as acme-client

### ISSUE THREAD

**Issue by nicovd737** [grindsa/acme2certifier#107]
Hello,
First thanks for your development, it's seems a very good idea and this can help us a lot.
We try to implement acme2certifier on docker and to make it work with Traefik.
All seems to work (traefik is communicating with acme2certifier and acme2certifier receive the communication from Traefik.
The only issue we have is that it seems that acme2certifier seems to ignore the file acme_srv.cfg.
We have configured it to communication with a Windows Root CA following your communication but we don't see any communication.
By the way we have tried to access the Root CA from the acme container but and it's working (with curl).

Here is our configuration:

cat acme_srv.cfg 
[DEFAULT]
debug: true

[CAhandler]
 CA specific options
handler_file: /var/www/acme2certifier/volume/ca_handler.py
host: rootca.server
user: Admin
password: xxxxxxx
ca_bundle: root-ca.pem
auth_method: basic
template: "Web-2023"

[DBhandler]
#dbfile: /var/lib/acme/db.sqlite3

[Certificate]
revocation_reason_check_disable: False

[Order]
tnauthlist_support: False

We have copied the file mscertsrv_ca_handler.py into ca_handler.py
When we chec the logs, we can see lot of activity but nothing regarding our Root CA server.

In the traefik logs, we can see the below error:

time="2023-05-23T13:28:40Z" level=error msg="Unable to obtain ACME certificate for domains \"piwigo.xxx.xxx\": unable to generate a certificate for the domains [piwigo.xxx.xxx]: error: one or more domains had a problem:\n[piwigo.xxx.xxx] acme: error: 403 :: POST :: https://dokmgt01:22443/acme/authz/ :: urn:ietf:params:acme:error:unauthorized :: authorizations lookup failed\n" providerName=myresolver.acme ACME CA="https://dokmgt01:22443/directory" routerName=piwigo@docker rule="Host(`piwigo.xxx.xxx`)"

the server "dokmgt01" is the docker server where the acme2certifier is running.

I assume that the problem is somewhere between acme2certifier container and our Windows root CA server.

We are a bit lost.
Any help would be very appreciated.
Many thanks
Nico

**Comment by nicovd737** [comment]
We have also this error in acme2certifier logs:

Error.enrich_error()
Error.acme_errormessage(urn:ietf:params:acme:error:unauthorized)
Authorization.new_post() returns: {"code": 403, "header": {}, "data": {"status": 403, "type": "urn:ietf:params:acme:error:unauthorized", "detail": "authorizations lookup failed"}}
10.0.0.2 /acme/authz/ {'code': 403, 'header': {}, 'data': {'status': 403, 'type': 'urn:ietf:params:acme:error:unauthorized', 'detail': 'authorizations lookup failed'}}
[pid: 30|app: 0|req: 168/267] 10.0.0.2 () {38 vars in 551 bytes} [Tue May 23 15:28:39 2023] POST /acme/authz/ => generated 108 bytes in 10 msecs (HTTP/1.1 403) 1 headers in 66 bytes (2 switches on core 0)

**Comment by nicovd737** [comment]
Hello,
Thanks for your reply.
Ok i could send the debug by mail and also share a pyc file which contains also some errors.
What's your mail adress ?
Thanks
Nico

**Comment by grindsa** [comment]
please send logs and pyc to grindelsack@gmail.com

**Comment by grindsa** [comment]
Thank you for sending the logs.  I would say that the challenge validations failed. It seems that traefik is using [tls-alpn](https://datatracker.ietf.org/doc/html/rfc8737) verification which means that acme2certifier will connect to `piwigo.xxx.xxx`  and expects a self-signed certificate to be presented which: includes:
-  an SAN matching  `piwingo.xxx.xxx`
-  a certificate extension containing an sha-256 digest

What i see from the logs is that a your traefik instance is presenting a certificate but with no SAN. Thus, validation fails and you see the error. Below related entries from your logs:
```
Challenge._validate_alpn_challenge(IOsVYikfwGW0:piwigo.payerne.local:6wUm0OEyusep3gxH0fnF6hU3SgSqfwTq)
fqdn_resolve() ended with: 10.203.2.228/False
sha256_hash_hex()
sha256_hash_hex() ended with 7a20112c4983e7882f02e09a76797c6a0042f6baae050473bdaca4024c66344b
b64_encode()
computed value: BCB6IBEsSYPniC8C4Jp2eXxqAEL2uq4FBHO9rKQCTGY0Sw==
servercert_get(piwigo.payerne.local:443)
servercert_get() configure proxy: piwigo.payerne.local:443 version: TLSv1.3
Challenge._extensions_validate(BCB6IBEsSYPniC8C4Jp2eXxqAEL2uq4FBHO9rKQCTGY0Sw==/piwigo.payerne.local)
cert_san_get()
cert_san_get() ended
fqdn_in_san_check()
fqdn_in_san_check() ended with: False
fqdn check against san failed
Challenge._extensions_validate() ended with: False
Challenge._validate_alpn_challenge() ended with: False/False
```

I am not sure if the prolblem is with Traeffik or with a2c, thus i am planning to setup a lab over weekend and investigate more.

Do you have the option to change the challenge-verification type to [http-01[(https://doc.traefik.io/traefik/https/acme/)?

if you want to test the conenction to your CA by disabling challenge validation but this is a severe security issue and i suggest to do this for testing only
```
[Challenge]
# when true disable challenge validation. Challenge will be set to 'valid' without further checking
# THIS IS A SEVERE SECURTIY ISSUE! Please do only for testing/debugging purposes
challenge_validation_disable: True
```

**Comment by grindsa** [comment]
any chance to get the docker compose yaml file? This would ease things on my side....

**Comment by grindsa** [comment]
Just to sum up. There was indeed a bug in acme2certfier during tls-alpn-01 challenge validation. According to [RFC 8737](https://datatracker.ietf.org/doc/html/rfc8737) ... `the ACME server MUST provide an ALPN extension with the single protocol name "acme-tls/1" and an SNI extension containing only the domain name being validated during the TLS handshake.` ... It seems that traefik takes routing decisions based on the presence of the ALPN extension which was not set by a2c and hence routing and challenge validation failed.  This issue has been addressed in 14e4cb0d78becc9c2a7c71c73a8ae542b6c2c759 and correction has been verified with a [new traefik regression workflow](https://github.com/grindsa/acme2certifier/commit/f334774ae133e528fe5aa5f350fbebe677481b15). 

I am planning to release a new version during the weekend; there are some smaller things which needs to be addressed before.

**Comment by grindsa** [comment]
closed as fix got included in v0.27

**Comment by nicovd737** [comment]
Thanks a lot for your support !!
